src/mono/netcore/System.Private.CoreLib/src/System/RuntimeType.Mono.cs

```diff
@@ -2454,12 +2454,20 @@ private RuntimeType[] GetNestedTypes_internal(string? displayName, BindingFlags
                 for (int i = 0; i < n; i++)
                 {
                     var th = new RuntimeTypeHandle(h[i]);
-                    a[i] = (RuntimeType)GetTypeFromHandle(th);
+                    a[i] = GetTypeFromHandle(th);
                 }
                 return a;
             }
         }
 
+        public static new RuntimeType GetTypeFromHandle(RuntimeTypeHandle handle)
```

Comment:
Some of these are going to break analysis within IL Linker whenever they're encountered. IL Linker doesn't hardcode runtime implementation details such as these optimizations - we wouldn't want to hardcode it into linker. Linker operates on public APIs only.

E.g. this one is normally [an intrinsic](https://github.com/mono/linker/blob/01f8592b4dcbc5a5e5e395cc2dba31260e3d3114/src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs#L263-L264), so is e.g. `EmptyTypes`. This will make it more difficult for IL Linker to statically analyze what's going on in these spots.

Cc @vitek-karas

Reply:
🤔 The covariant returns allow a lot of the casts back to `RuntimeType` to be dropped (haven't applied all the changes in dropping the casts); and also to devirtualized as direct calls as its a sealed type, rather than virtual calls via `Type`

Reply:
Do you have any perf numbers for this?

The cast should be pretty cheap because it's an exact match (the cast will trivially succeed in the part of the type check that RyuJIT typically inlines - it succeeds before actually calling into the casting logic).

Reply:
Type checks then might not mean much?

![image](https://user-images.githubusercontent.com/1142958/100634781-9f0b6f80-3327-11eb-929a-b8033a645ba8.png)

Though it does also drop the virtual calls:

![image](https://user-images.githubusercontent.com/1142958/100635270-3d97d080-3328-11eb-8f4c-df5f03c39090.png)

Reply:
I meant the non-R2R case: https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgZgARRwQA2cAJnlgEx4DCeA3gLABQe7l+WAjAGx14AsgAoA9sABWcAMYw8ogJSNWHVZQDseYbQWiA3Co4BfVkaA

RyuJIT generates two checks before calling the casting helper - is it a null object, or is it exactly the same type that we're casting to. If yes, it skips calling into runtime completely.

This change is still an improvement, but is it an improvement that would justify likely having to implement & maintain a IL linker feature? E.g. [just earlier today](https://github.com/mono/linker/pull/1655) I implemented intrinsic recognition of calls to `Type.UnderlyingSystemType` within linker motivated by the use within CoreLib - after this change, any call to get_UnderlyingSystemType on a RuntimeType would no longer be intrinsically recognized without linker work because this is a callvirt to a different method.